discovery_engine_core/core/src/mab.rs

```diff
@@ -46,108 +46,102 @@ impl BetaSample for BetaSampler {
 pub(crate) trait Bucket<T> {
     /// Returns the alpha parameter of the beta distribution.
     fn alpha(&self) -> f32;
+
     /// Returns the beta parameter of the beta distribution.
     fn beta(&self) -> f32;
-    /// Returns `true` if the bucket contains no elements.
+
+    /// Checks if the bucket is empty.
     fn is_empty(&self) -> bool;
+
     /// Removes the next best element from this bucket and returns it, or `None` if it is empty.
     fn pop(&mut self) -> Option<T>;
 }
 
-fn pull_arms<B, T>(beta_sampler: &impl BetaSample, buckets: &mut [&mut B]) -> Result<T, Error>
+impl<B, T> Bucket<T> for &mut B
 where
     B: Bucket<T>,
 {
-    let sample_from_bucket = |bucket: &B| beta_sampler.sample(bucket.alpha(), bucket.beta());
-
-    let mut buckets = buckets.iter_mut();
-
-    let first_bucket = buckets.next().ok_or(Error::NoBucketsToPull)?;
-    let first_sample = sample_from_bucket(first_bucket)?;
-
-    let bucket = buckets
-        .try_fold(
-            (first_sample, first_bucket),
-            |max, bucket| -> Result<_, Error> {
-                let sample = sample_from_bucket(bucket)?;
-                if let Ordering::Greater = nan_safe_f32_cmp(&sample, &max.0) {
-                    Ok((sample, bucket))
-                } else {
-                    Ok(max)
-                }
-            },
-        )?
-        .1;
-
-    bucket.pop().ok_or(Error::EmptyBucket)
+    fn alpha(&self) -> f32 {
+        (&**self).alpha()
```

Comment:
would also work. same for `beta` and `is_emtpy`
```suggestion
        (**self).alpha()
```